### ISSUE TITLE
notebook installs

### ISSUE THREAD

**Issue by jeremymanning** [ContextLab/data-wrangler#20]
when installing davos in notebooks, it requires restarting the runtime.

@paxtonfitzpatrick would it be possible to use your "davos" trick when installing in notebooks?  i think it'd entail something like the following (on installation):

1.) determine we're installing from a notebook environment
2.) if so, automatically restart the runtime

i wonder if we could even include davos as a dependency of pydata-wrangler and just call the smuggle function directly... 🤔

**Comment by paxtonfitzpatrick** [comment]
@jeremymanning can you post a bit more info on what's going on? Namely:
- the command you're running
- the traceback/output it generates
- whether this is happening in a Jupyter or Colab notebook
   - if Jupyter, your `IPython.__version__` and `ipykernel.__version__`

Also:

> when installing davos in notebooks, it requires restarting the runtime.

Is this an issue with installing `davos` or `pydata-wrangler`?

**Comment by jeremymanning** [comment]
@paxtonfitzpatrick sorry-- i meant when you install *datawrangler* (not davos) in notebooks, it requires a runtime restart.  it's definitely a datawrangler "issue" (although it's related to how datawrangler's requirements differ from colab's built in libraries, not something in datawrangler per se).  to reproduce, just open a new colab notebook and run:

```
!pip install pydata-wrangler
```

You'll get a "restart runtime" button and datawrangler won't work correctly until you restart the runtime.  I'm just wondering if there's a more elegant solution for this, whereby installing pydata-wrangler restarts the runtime automatically under these circumstances, rather than requiring a manual restart.

**Comment by paxtonfitzpatrick** [comment]
Got it 👍 

**tl;dr**:
1. There's no (viable) way to automatically restart the Colab runtime
2. Using `davos` to `smuggle datawrangler` works without requiring you to restart the runtime
2. I don't think it's possible to build this functionality into `pydata-wrangler` itself, either natively or by requiring and using `davos`.
4. I think the best solution is just to use `davos` directly and `smuggle datawrangler` whenever you're working in a Colab notebook (this won't affect Jupyter or non-interactive Python, plus this is one of the things `davos` was designed to do anyway).  It might be worth adding a note to the docs that suggests using `davos` when working in Colaboratory.
5. I found a bug in `davos` while working on this, so this won't work immediately, but it should only take me <1 hour to fix. I'll merge into the dev branch of ContextLab/davos and tag this issue in the PR when I'm done.

> I'm just wondering if there's a more elegant solution for this, whereby installing pydata-wrangler restarts the runtime automatically under these circumstances, rather than requiring a manual restart.
- So, it is _possible_ to restart a Colab runtime from Python, but there's no "clean" way to do it like you can in a Jupyter notebook. The way [`davos`](https://github.com/ContextLab/davos) does this for Jupyter involves accessing the notebook JavaScript object, which Colaboratory actively blocks you from doing from Python.  The one way you can get the Colab runtime to restart is to kill the Python interpreter process, e.g.:
   ```python
   import os
   os.kill(os.getpid(), 9)    # any signal works
   ```
   However, killing the runtime this way prompts a series of pop-ups which can't be suppressed, including this one:
   ![Screen Shot 2021-07-21 at 2 47 09 PM](https://user-images.githubusercontent.com/26118297/126543101-27350b18-8a81-4ea5-a8eb-a7e95d1dfca7.png)
   It also causes a bug where the "Files" viewer pane shows this, even after the runtime reconnects:
   ![Screen Shot 2021-07-21 at 2 51 20 PM](https://user-images.githubusercontent.com/26118297/126543627-bb651a97-3300-4bec-b93f-18217ba310b8.png)
   So you end up unable to browse the file system until you factory reset the runtime, manually.  IMO this ultimately this isn't viable, so `davos` doesn't support automatically restarting in Colab.


- The good news is using installing & importing the package with `davos` works without having to restart the runtime:
   ```python
   import davos
   smuggle datawrangler as dw    # pip: pydata-wrangler
   ```
   Both Colab (when you run a `!pip install` command), and `davos` (when you `smuggle` something) check newly installed packages against those already imported pretty much the same way.  The reason this works through `davos` but not Colab, however, is that Colab immediately issues that "Restart Runtime" warning message if any matches are found.  `davos`, however, will try to unload the old version of each package, and recursively replace all of its modules with those from the just-installed package.  This works for the vast majority of packages, and pretty much only fails when A) the package contains C extension modules *and* B) those modules were changed between the existing and smuggled versions.

   The two modules causing the warning for `datawrangler` are `matplotlib` and `mpl_toolkits`, both of which **seem** to be able to be reloaded without any issues. The only potential source of trouble with this that I can think of is that I know both IPython and Colab run some matplotlib-specific startup hooks, but iirc those do some pretty standard things that aren't super likely to change, except maybe across major versions. But I haven't played around with matplotlib after smuggling it, so I'm not 100% certain.

> i wonder if we could even include davos as a dependency of pydata-wrangler and just call the smuggle function directly
- You could, and you could set it up so that `matplotlib` and `mpl_toolkits` get reoladed when you `import datawrangler`, but unfortunately A) you couldn't do it using `davos` and would have to reimplement most of that package in `pydata-wrangler` to make this work, and B) this wouldn't get rid of the big red warning & button, because that's issued when the package is _installed_, rather than imported.  I also don't think it's possible to write this as a `setuptools` post-install class in `setup.py`, because the installation happens in a subprocess which wouldn't be able to view the notebook process's `sys.modules` anyway.
